src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs

```diff
@@ -3,273 +3,183 @@
-                Subject = clientId,
-                JwtIdentifier = Guid.NewGuid().ToString()
-            };
+            _clientId = clientId;
+            _audience = audience;
         }
 
-        public JsonWebToken(ICryptographyManager cryptographyManager, string clientId, string audience, IDictionary<string, string> claimsToSign, bool appendDefaultClaims = false)
-            : this(cryptographyManager, clientId, audience)
+        public JsonWebToken(
+             ICryptographyManager cryptographyManager,
+             string clientId,
+             string audience,
+             IDictionary<string, string> claimsToSign,
+             bool appendDefaultClaims = false)
+         : this(cryptographyManager, clientId, audience)
         {
-            ClaimsToSign = claimsToSign;
+            _claimsToSign = claimsToSign;
             _appendDefaultClaims = appendDefaultClaims;
         }
 
-        public string Sign(X509Certificate2 certificate, string base64EncodedThumbprint, bool sendX5C)
+        private string CreateJsonPayload()
         {
-            // Base64Url encoded header and claims
-            string token = Encode(certificate, base64EncodedThumbprint, sendX5C);
+            long validFrom = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
+            long validTo = validFrom + JwtToAadLifetimeInSeconds; // 10 min
 
-            // Length check before sign
-            if (MaxTokenLength < token.Length)
+            if (_claimsToSign == null || _claimsToSign.Count == 0)
             {
-                throw new MsalException(MsalError.EncodedTokenTooLong);
+                return $$"""{"aud":"{{_audience}}","iss":"{{_clientId}}","sub":"{{_clientId}}","nbf":"{{validFrom}}","exp":"{{validTo}}","jti":"{{Guid.NewGuid()}}"}""";
             }
 
-            byte[] signature = _cryptographyManager.SignWithCertificate(token, certificate);
-            return string.Concat(token, ".", UrlEncodeSegment(signature));
-        }
+            // extra claims
+            StringBuilder payload = new StringBuilder();
 
-        private static string EncodeSegment(string segment)
-        {
-            return UrlEncodeSegment(Encoding.UTF8.GetBytes(segment));
-        }
+            if (_appendDefaultClaims)
+            {
+                string defaultClaims = $$"""{"aud":"{{_audience}}","iss":"{{_clientId}}","sub":"{{_clientId}}","nbf":"{{validFrom}}","exp":"{{validTo}}","jti":"{{Guid.NewGuid()}}",""";
+                payload.Append(defaultClaims);
+            }
+            else
+            {
+                payload.Append("{");
+            }
 
-        private static string UrlEncodeSegment(byte[] segment)
-        {
-            return Base64UrlHelpers.Encode(segment);
-        }
+            int i = 0;
+            foreach (var kvp in _claimsToSign)
+            {
+                payload.Append($"\"{kvp.Key}\":\"{kvp.Value}\"");
+                if (i!= _claimsToSign.Count-1)
```

Comment:
might be cleaner to do 
```
if (kvp.Key != _claimsToSign.Last.Key)
{
    payload.Append(",");
}
```
Then you can get rid of the i counter.

Reply:
Last is part of Linq, and I think in this case it'll lead to the enumeration of the entire collection.